### PR TITLE
Create a MANDATORY Resource instance containing mandatory attributes. Merge that into the default resource.

### LINKED ISSUE #2298: Always use a fallback service.name in the Resource, if none is provided.
With https://github.com/open-telemetry/opentelemetry-specification/pull/1269 merged, we need to implement it.

### PR DESCRIPTION
**Pull request by jkwatson**
resolves #2298 in a less drastic way, which seems to be what the spec is being updated to require.

compare to #2310 

spec PR reference: https://github.com/open-telemetry/opentelemetry-specification/pull/1294

### CODE REVIEW

File: sdk/metrics/src/test/java/io/opentelemetry/sdk/metrics/spi/SdkMeterProviderFactoryTest.java
Comment (anuraaga) [inline sdk/metrics/src/test/java/io/opentelemetry/sdk/metrics/spi/SdkMeterProviderFactoryTest.java:27]: Think it's good if at least one of the tests has another attribute in it.

File: sdk/trace/src/test/java/io/opentelemetry/sdk/trace/SdkTracerProviderTest.java
Comment (anuraaga) [inline sdk/trace/src/test/java/io/opentelemetry/sdk/trace/SdkTracerProviderTest.java:48]: Do we need this mock?

**Review by anuraaga — APPROVED** [review]
Much cooler

File: sdk/trace/src/test/java/io/opentelemetry/sdk/trace/SdkTracerProviderTest.java
Comment (jkwatson) [reply sdk/trace/src/test/java/io/opentelemetry/sdk/trace/SdkTracerProviderTest.java:48]: probably not. Can just use `Resource.getEmpty()`, for sure.

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/SdkMeterProviderBuilder.java
Comment (Oberon00) [inline sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/SdkMeterProviderBuilder.java:59]: This is how I initially proposed the spec PR, but not how I reworked it shortly after and how it got approved 😃 
Now the spec basically says that we should treat the service.name exactly the same as we already treat the telemetry.sdk.name: It's there by default, but if the user set's their own resource explicitly, they need to merge it in manually.

See https://github.com/open-telemetry/opentelemetry-specification/pull/1294#issuecomment-747589223 and https://github.com/open-telemetry/opentelemetry-specification/blob/39fa82f0a7e58975782520c304768d63da0c7612/specification/resource/sdk.md (not yet merged).

If you think the first version of the PR was better, you should say so on the spec. 

@anuraaga @jkwatson You approved the spec PR after I made this change, but maybe you want to check again?

**Review by Oberon00 — CHANGES_REQUESTED** [review]

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/SdkMeterProviderBuilder.java
Comment (jkwatson) [reply sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/SdkMeterProviderBuilder.java:59]: I removed my approval, because I was mostly going on the PR description, which seems to match what I've implemented here. What's now written in the spec is no longer comprehensible to me. :(

Comment (Oberon00) [reply sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/SdkMeterProviderBuilder.java:59]: @jkwatson To really dismiss your approval you have to expand the box at the bottom of the PR and click the dots next to your approval there (or add a request-changes review). Adding a comment-only review does not remove the approval.

**Comment by jkwatson** [comment]
@Oberon00 Do you think this PR is now inline with the current state of your in-flight spec PR?  https://github.com/open-telemetry/opentelemetry-specification/pull/1294

**Comment by jkwatson** [comment]
this is dependent on #2542 to get the merging to happen in the right order

File: sdk/common/src/main/java/io/opentelemetry/sdk/resources/Resource.java
Comment (Oberon00) [inline sdk/common/src/main/java/io/opentelemetry/sdk/resources/Resource.java:62]: I don't think this needs to be separated from DEFAULT. Also, this would be the only `public` field in this class, the others are `private` and have getters.

**Review by Oberon00 — APPROVED** [review]

File: sdk/common/src/main/java/io/opentelemetry/sdk/resources/Resource.java
Comment (jkwatson) [reply sdk/common/src/main/java/io/opentelemetry/sdk/resources/Resource.java:62]: I like having it be a separate thing in the code, but agree it should be private, and can be accessed via the default Resource.

Comment (jkwatson) [reply sdk/common/src/main/java/io/opentelemetry/sdk/resources/Resource.java]: done

**Pull request merged**
